compiler/passes/denormalize.cpp

```diff
@@ -669,3 +676,259 @@ inline bool unsafeExprInBetween(Expr* e1, Expr* e2, Expr* exprToMove,
+}
+
+/////////////////////////////////////////////////////////////////////////////
+// collapseTrivialMoves() converts:
+//   move(source, expr)  // move1
+//   move(dest, source)  // move2
+// provided:
+//   'move1' and 'move2' are adjacent or have only DefExprs in between
+//   'source' has no other references
+// to:
+//   move(dest, expr)
+
+// In some cases codegen adds an appropriate dereference, widening, etc.
+// for a symbol-to-symbol move and not for a call-to-symbol move.
+// So collapse only moves for which such additions are not needed.
+static bool okSymbol(Symbol* sym) {
+  return sym->qual == QUAL_VAL || sym->qual == QUAL_CONST_VAL;
+}
+
+static bool okToCollapse(SymExpr* dest, SymExpr* source) {
+  Symbol *destSym = dest->symbol(), *sourceSym = source->symbol();
+  return
+    destSym->type == sourceSym->type        &&
+    ! sourceSym->hasFlag(FLAG_CONFIG)       &&
+    ! sourceSym->hasFlag(FLAG_EXPORT)       &&
+    ! sourceSym->hasFlag(FLAG_EXTERN)       &&
+    okSymbol(destSym) && okSymbol(sourceSym);
+}
+
+static bool closeEnough(Expr* move1, Expr* move2) {
+  Expr* curr = move1;
+  for (int dist = 0; dist < 5; dist++) {  // heuristically allow <=5 defs
+    curr = curr->next;
+    if (curr == move2) return true;
+    if (!curr || ! isDefExpr(curr)) return false;
+  }
+  return false;
+}
+
+// Returns 'move1', or NULL if the desired pattern is not present.
+// 'sourceSE' is the SymExpr for 'source' in 'move2'.
+static CallExpr* singleMoveTo(CallExpr* move2, SymExpr* sourceSE) {
+  Symbol* source = sourceSE->symbol();
+  SymExpr* otherSE = source->firstSymExpr();
+  // Specialize for exactly two references to 'source'.
+  if (otherSE == sourceSE) {
+    // need source -> sourceSE -> otherSE -> NULL
+    otherSE = otherSE->symbolSymExprsNext;
+    if (otherSE == nullptr || otherSE->symbolSymExprsNext != nullptr)
+      return nullptr; // 1 or >2 references
+  }
+  else {
+    // need source -> otherSE -> sourceSE -> NULL
+    if (otherSE->symbolSymExprsNext != sourceSE ||
+        sourceSE->symbolSymExprsNext != nullptr)
+      return nullptr; // >2 references
+  }
+
+  CallExpr* move1 = toCallExpr(otherSE->parentExpr);
+  if (move1 != nullptr                                    &&
+      otherSE == move1->get(1)                            &&
+      move1->isPrimitive(PRIM_MOVE)                       && 
+      (move2 == move1->next || closeEnough(move1, move2)) )
+    return move1;
+  else
+    return nullptr;
+}
+
+static void collapseTrivialMoves() {
+  // Empirically, running collapseTrivialMoves() the second time
+  // would not result in any additional removals. So, do it just once.
+  //
+  // This work could be done on a per-function basis using collectCallExprs().
+  // Simply traversing 'gCallExprs' avoids the overhead of collectCallExprs().
+  for_alive_in_Vec(CallExpr, move2, gCallExprs) {
+   if (move2->isPrimitive(PRIM_MOVE))
+    if (SymExpr* dest = toSymExpr(move2->get(1)))
+     while (true) {
```

Comment:
Could we adjust this to have it be a `bool` flag instead like `checkForMoves`? I always get nervous when I see `while (true)`.